src/QirRuntime/public/QuantumApiBase.hpp

```diff
@@ -0,0 +1,260 @@
+#pragma once
+
+#include <exception>
+
+#include "IQuantumApi.hpp"
+
+namespace quantum
+{
+    class CQuantumApiBase : public IQuantumApi
+    {
+      public:
+        // Qubit management
+        Qubit AllocateQubit() override
+        {
+            throw std::logic_error("not_implemented");
+        }
+        void ReleaseQubit(Qubit qubit) override
+        {
+            throw std::logic_error("not_implemented");
+        }
+        void GetState(TGetStateCallback callback) override
+        {
+            throw std::logic_error("not_implemented");
+        }
+        virtual std::string DumpQubit(Qubit qubit) override
+        {
+            throw std::logic_error("not_implemented");
+        }
+
+        // Shortcuts
+        void CX(Qubit control, Qubit target) override

```

Comment:
I'm still a little confused why these shortcut methods exist, and why they are decomposed here instead of in some other place. Long term, I think we'd want to see all decompositions removed in favor of pure Q# decomposition in the targeting package (see my other comment). For now, I think I would be more comfortable with these if they were moved to the `CCanonicalDecomposer` class so that all the decompositions are in the same place and don't start spreading out. That way, this base class has purely "not implemented" logic_error versions of every function and be a truly common base that can be used by any later implementation.

Reply:
I've been asked to follow the incubation IQuantumApi as close as possible and it happened to differentiate between singly versus multi-controlled operations. @avasch01 would have more context.

Reply:
I think the API itself is good (especially considering I can already see think of a few places where specific targets would want to use these over the multi-controlled version), I just disagree with having the decomposition live in this particular base class. I'm just in favor of having a base class where every method is "not implemented" and leaving the implementation of any supported methods to the child classes. But this is not a big deal for this PR and the pattern of where decompositions live and how they imply what implementations are present and native to the target is something that we will need to do more review and design on later anyway.

For context, over in the feature/decomp branch I ended up needing to define an internal, singly-controlled variant for `X` and `Z` as well as versions that did not support any controls (see [ApplyControlledX](https://github.com/microsoft/qsharp-runtime/blob/feature/decomp/src/Simulation/TargetDefinitions/Intrinsic/ApplyControlledX.qs) as an example). I would've called this `CX` except that this is already taken by a library operation, [Microsoft.Quantum.Canon.CX](https://github.com/microsoft/QuantumLibraries/blob/b802b254b7cb98049573a367ed399e23ad784e59/Standard/src/Canon/CommonGates.qs#L38). As we figure out how to integrate the `IQuantumAPI` with the target packages and the runtime, it may mean moving some of these callables out of the libraries and into the runtime to allow them to be treated as target-overridable intrinsics and redefining the operations I put into these first target packages to make them line up with the simulator API (possibly making them public instead of internal in the process). But all that is a much bigger cross-repo change that we would need to coordinate later. Suffice it to say, I think the API design is good, I just think it implies the need for this larger design update and refactor at some point in the future :).